pkg/storage/etcd3/etcdwatcher/watcher.go

```diff
@@ -0,0 +1,340 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdwatcher
+
+import (
+	"container/list"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/coreos/etcd/clientv3"
+	etcdrpc "github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	"golang.org/x/net/context"
+)
+
+type Watcher interface {
+	// Watch watches on a key and returns a watch.Interface that transfer relevant notifications.
+	// If rev is zero, it will try to get existing objects if any and then watch after that revision.
+	// If rev is non-zero, it will watch from given revision.
+	// If recursive is false, it watches on given key.
+	// If recursive is true, it watches any children and directories under the key, excluding the root key itself.
+	// filter must be non-nil. Only if filter returns true will the changes be returned.
+	Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error)
+}
+
+// New returns a Watcher.
+func New(client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) Watcher {
+	return newWatcher(client, codec, versioner)
+}
+
+// watcher implements Watcher.
+type watcher struct {
+	client    *clientv3.Client
+	codec     runtime.Codec
+	versioner storage.Versioner
+}
+
+// watchChan implements watch.Interface.
+type watchChan struct {
+	watcher       *watcher
+	key           string
+	initialRev    int64
+	recursive     bool
+	filter        storage.FilterFunc
+	ctx           context.Context
+	cancel        context.CancelFunc
+	resultChan    chan watch.Event
+	eventChan     chan *event
+	resultQueue   *list.List
+	errChan       chan error
+	waitSync      chan struct{}
+	waitEventProc sync.WaitGroup
+}
+
+func newWatcher(client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) *watcher {
+	return &watcher{
+		client:    client,
+		codec:     codec,
+		versioner: versioner,
+	}
+}
+
+func (w *watcher) Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error) {
+	if recursive && !strings.HasSuffix(key, "/") {
+		key += "/"
+	}
+	wc := &watchChan{
+		watcher:     w,
+		key:         key,
+		initialRev:  rev,
+		recursive:   recursive,
+		filter:      filter,
+		resultChan:  make(chan watch.Event, 1),
+		eventChan:   make(chan *event, 100),
+		resultQueue: list.New(),
+		errChan:     make(chan error, 1),
+		waitSync:    make(chan struct{}),
+	}
+	wc.ctx, wc.cancel = context.WithCancel(ctx)
+	go wc.run()
```

Comment:
This is already in Watch() call.

Let me clarify just in case. The pattern is a little different:
- I have a watcher which provides Watch(), recursive or not.
- Watch() returns a watchChan, which will be used to return from storage.Interface.Watch()/WatchList().

rev=0 to get current state was originally designed to clean the code (#846). I think they are replaced by informers afterwards. I might want to deprecate this behavior if possible.